### PR TITLE
ALTAPPS-669: Shared base study plan UI

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-669](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-669)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
~~Sentry Performance Monitoring of screen loading is added for new screens~~
- [x] View analytics events are added for new screens;
- [x] New analytics events are documented;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Change internal activities structure from  map of `SeticionId -> Set<LearningActivity>` to map of `ActivityId -> LearningActivity` to access activity by id;
- Add track fetching to show track tittle;
- Update `SectionClicked` message;
- Add `ActivityClicked` message;
- Add section and activities details to viewState

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/injection/StudyPlanComponentImpl.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package org.hyperskill.app.study_plan.widget.injection
+
+import org.hyperskill.app.core.injection.AppGraph
+import org.hyperskill.app.core.presentation.ActionDispatcherOptions
+import org.hyperskill.app.study_plan.widget.presentation.StudyPlanWidgetActionDispatcher
+import org.hyperskill.app.study_plan.widget.presentation.StudyPlanWidgetReducer
+import org.hyperskill.app.study_plan.widget.view.StudyPlanWidgetViewStateMapper
+
+class StudyPlanComponentImpl(private val appGraph: AppGraph) : StudyPlanWidgetComponent {
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/injection/StudyPlanComponentImpl.kt:9]: I think we should rename `StudyPlanComponentImpl -> StudyPlanWidgetComponentImpl`

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenActionDispatcher.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import org.hyperskill.app.core.presentation.ActionDispatcherOptions
+import ru.nobird.app.presentation.redux.dispatcher.CoroutineActionDispatcher
+
+class StudyPlanScreenActionDispatcher(
+    config: ActionDispatcherOptions,
+    private val analyticInteractor: AnalyticInteractor
+) : CoroutineActionDispatcher<StudyPlanScreenFeature.Action, StudyPlanScreenFeature.Message>(config.createConfig()) {
+    override suspend fun doSuspendableAction(action: StudyPlanScreenFeature.Action) {
+        when (action) {
+            is StudyPlanScreenFeature.InternalAction.LogAnalyticEvent -> {
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenActionDispatcher.kt:13]: Why do we need `InternalAction` interface if we don't use `when` on it?

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenReducer.kt
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenReducer.kt:1]: That's why I don't like approach with composed features - so many boilerplate

**Review by vladkash — CHANGES_REQUESTED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenReducer.kt
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenReducer.kt:1]: Yes, you are right. We have a boilerplate.
But on the other hand, you are free to compose features like you want. You don't need to rewrite existing code, only write new code.

**Review by vladkash — APPROVED** [review]
Maybe we should wait for @ivan-magda review before merging

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenFeature.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+object StudyPlanScreenFeature {
+
+    fun isAvailable(buildKonfig: BuildKonfig): Boolean =
+        BuildKonfig.IS_INTERNAL_TESTING ?: (buildKonfig.buildVariant == BuildVariant.DEBUG)
+
+    internal data class State(
+        val toolbarState: GamificationToolbarFeature.State,
+        val studyPlanWidgetState: StudyPlanWidgetFeature.State
+    )
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenFeature.kt:17]: I don't like this, in my opinion, it's better to have explicit and linear execution with some boilerplate

**Comment by ivan-magda** [comment]
> Maybe we should wait for @ivan-magda review before merging

@XanderZhu merge without my review, I'll take a look at the final feature PR

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenFeature.kt [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package org.hyperskill.app.study_plan.screen.presentation
+
+import org.hyperskill.app.analytic.domain.model.AnalyticEvent
+import org.hyperskill.app.config.BuildKonfig
+import org.hyperskill.app.core.domain.BuildVariant
+import org.hyperskill.app.gamification_toolbar.presentation.GamificationToolbarFeature
+import org.hyperskill.app.study_plan.widget.presentation.StudyPlanWidgetFeature
+
+object StudyPlanScreenFeature {
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenFeature.kt:9]: 1. Why do we have several study plan features: `StudyPlanScreenFeature` and `StudyPlanWidgetFeature`?
2. What purpose of the `StudyPlanScreenFeature`?
3. What purpose of the `StudyPlanWidgetFeature`?

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/screen/presentation/StudyPlanScreenFeature.kt:9]: @ivan-magda 

`StudyPlanWidgetFeature` is a widget that contains the logic of `StudyPlan`: how to load `StudyPlan`, how to handle user interaction with it, etc.

`StudyPlanScreenFeature` is a screen feature, which is just a host for other features.
It sends its own view analytics event and could delegate screen events such as `SwipeRefresh` to child features.
It is just a box for other features, e.g. SudyPlanWidget, GamificationToolbar, or ProblemsLimit.

The reason, why I separated `StudyPlanWidgetFeature`  and `StudyPlanScreenFeature` is to not mix complex `StudyPlanWidgetFeature.State` with other features states.

**Pull request merged**